core/src/main/scala/cats/data/Continuation.scala

```diff
@@ -0,0 +1,108 @@
+package cats
+package data
+
+import java.io.Serializable
+import scala.annotation.tailrec
+
+sealed abstract class Continuation[O, +I] extends Serializable {
+  def map[I2](fn: I => I2): Continuation[O, I2] =
+    Continuation.Mapped(this, fn)
+
+  def flatMap[I2](fn: I => Continuation[O, I2]): Continuation[O, I2] =
+    Continuation.FlatMapped(this, fn)
+
+  final def apply(fn: I => O): O = {
+    import Continuation._
+    val re = new RunEnv[O]
+    re.eval(this, re.ScalaFn(fn)).get
+  }
+}
+
+object Continuation {
+  def pure[O] = new PureBuilder[O]
+  class PureBuilder[O] {
```

Comment:
`final` and `private[Continuation]` ? The former for sure, the latter mostly because I saw that's how we've been doing it: https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/OptionT.scala#L167